ssz-rs/src/merkleization/merkleize.rs

```diff
@@ -320,9 +449,89 @@ pub fn compute_merkle_tree(
         // NOTE: children.len() == 2 * BYTES_PER_CHUNK
         let (parent, children) = focus.split_at_mut(children_index);
         let (left, right) = children.split_at(BYTES_PER_CHUNK);
-        hash_nodes(hasher, left, right, &mut parent[..BYTES_PER_CHUNK]);
+        hash_nodes(left, right, &mut parent[..BYTES_PER_CHUNK]);
+    }
+}
+
+// Process a subtree of the Merkle tree.
+fn process_subtree(buffer: &mut [u8], size: usize) {
+    // No pairs to process
+    if size < 2 {
+        return;
+    }
+    // Process from leaves up, handling pairs of nodes
+    for i in (0..size - 1).rev().step_by(2) {
+        let left = &buffer[i * BYTES_PER_CHUNK..(i + 1) * BYTES_PER_CHUNK];
+        let right = &buffer[(i + 1) * BYTES_PER_CHUNK..(i + 2) * BYTES_PER_CHUNK];
+
+        // Check if both left and right are all zero or precomputed zero subtree hashes
+        let maybe_parent = zero_parent_level(left, right, &CONTEXT);
+        // Compute parent hash
+        let parent_hash = if let Some(zero_hash) = maybe_parent {
+            zero_hash
+        } else {
+            // Compute the hash normally
+            hash_chunks(left, right)
+        };
+
+        // Store at parent position (i/2)
+        // SAFETY: checked subtraction is unnecessary, as i >= 1; qed
+        let parent_index = i / 2;
+        let parent_slice =
+            &mut buffer[parent_index * BYTES_PER_CHUNK..(parent_index + 1) * BYTES_PER_CHUNK];
+        parent_slice.copy_from_slice(&parent_hash);
+    }
+}
+
+// Determines subtree level if the node is a zero subtree.
+fn zero_subtree_level(node: &[u8], context: &Context) -> Option<usize> {
+    // Check against precomputed zero hashes in the context
+    let total_levels = MAX_MERKLE_TREE_DEPTH;
+    for lvl in 0..total_levels {
+        if node == &context[lvl][..] {
+            return Some(lvl + 1);
+        }
+    }
+    None
+}
+
+// Determines the parent hash for a pair of nodes if they are both
+// zero subtrees, allowing hashing to be skipped.
+fn zero_parent_level(
+    left: &[u8],
+    right: &[u8],
+    context: &Context,
+) -> Option<[u8; BYTES_PER_CHUNK]> {
+    let left_level = zero_subtree_level(left, context)?;
+    let right_level = zero_subtree_level(right, context)?;
+    if left_level == right_level {
+        // Parent level is one higher
+        let parent_lvl = left_level + 1;
+        // Zero hashes in context start at index 0 for subtree level 1
+        Some(context[parent_lvl - 1].try_into().expect("valid subtree hash"))
+    } else {
+        None
+    }
+}
```

Comment:
I get this logic and it is a nice implementation, however I wonder we can make a smarter optimization by leveraging the fact that you know the generalised index of the element you want to prove.
With that information you can find out what are the required indexes to calculate the merkle proof and upon futher analysis selecting the precomputed zero hashes available.

Consider the following tree
```rs
//       0       <- Root (index 0)
//    /     \
//   1       2   <- Interior nodes
//  / \     / \
// 3   4   5   6 <- Leaf nodes
```
If you need to prove at index `3`, then you don't need to compute the right side of the tree at all and get the precomputed hash for index `2`. Does it make sense?

Reply:
With this I'm hinting that, especially in the context of producing a single proof, maybe the benefits of multi-threading can be surpassed by considering that you always don't need to compute at least half of the tree. Perhaps the two strategies can be combined in the context of multi-proving.

Reply:
> If you need to prove at index `3`, then you don't need to compute the right side of the tree at all and get the precomputed hash for index `2`. Does it make sense?

I think this depends how the API is going to look. Right now `compute_merkle_tree()` is not aware of the index and it's a general implementation that's going to be used for both prove and multiprove. I do wanna focus on making this fast first and then specialize, but I like your idea a lot!